### ISSUE TITLE
implementation of flip_gradient on dynet

### ISSUE THREAD

**Issue by iamyoungjo** [clab/dynet#318]
Hello.

This is more like proposal or suggestion than issue finding. I would like to implement flip_gradient, which is kind of a key feature for adversarial training. But dynet doesn't have it. I know this will never be a complicated tweak as it's just putting negative on backprop process only. But as I am lack of fundamental ideas of dynet structure, can you suggest which codes that I need to look into? I am trying to find the gradients in node unit.

I believe there could be many ways to make this work, but I need your suggestion.

Thanks,
YJ

**Comment by neubig** [comment]
Sure, for this it'd probably be best to create a "flip_gradient" operation. A brief description of how to do so is here at the very end of the slides:
http://phontron.com/slides/emnlp2016-dynet-tutorial-part1.pdf

My suggestion would be to take a look at the code for the "nobackprop" operation and the unary "-" operation, and combine the ideas together. If you get this working we'd love to have a pull request!

**Comment by iamyoungjo** [comment]
This tutorial is exactly what I was looking for. I will work on it and get back to you soon!

**Comment by iamyoungjo** [comment]
Hi,

I implemented as you suggested. 

This is a snippet from nodes.cc
```c
// Flip gradient For adversarial training
template<class MyDevice>
void FlipGradient::forward_dev_impl(const MyDevice & dev, const vector<const Tensor*>& xs, Tensor& fx) const {
  fx.d = xs[0]->d;
  fx.v = xs[0]->v;
}

template<class MyDevice>
void FlipGradient::backward_dev_impl(const MyDevice & dev,
                                   const vector<const Tensor*>& xs,
                                   const Tensor& fx,
                                   const Tensor& dEdf,
                                   unsigned i,
                                   Tensor& dEdxi) const {
  dEdxi.tvec().device(*dev.edevice) = dEdf.tvec().unaryExpr(scalar_negative_op<float>());
}
DYNET_NODE_INST_DEV_IMPL(FlipGradient)
```
And this is from simd_funtors.h. 
```c
namespace dynet {
template<typename Scalar> struct scalar_negative_op {
  EIGEN_EMPTY_STRUCT_CTOR(scalar_negative_op)
  DYNET_DEVICE_FUNC inline const Scalar operator() (const Scalar& x) const {
    return -x;
  }
  template <typename Packet>
  DYNET_DEVICE_FUNC inline Packet packetOp(const Packet& x) const {
    using namespace Eigen::internal;
    return pnegate(x);
  }
};
}

namespace Eigen { namespace internal {
template<typename Scalar>
struct functor_traits<dynet::scalar_negative_op<Scalar> > {
  enum {
    Cost = NumTraits<Scalar>::MulCost * 2,
    PacketAccess = packet_traits<Scalar>::HasNegate
  };
};
} }
```

Currently I am writing a sample code to see if this works fine in actual training process. If I missed some part to be changed or got anything wrong, glad to have your feedback.

Thanks,

**Comment by iamyoungjo** [comment]
Hi,

I implemented as you suggested. 

This is a snippet from nodes.cc
```c
// Flip gradient For adversarial training
template<class MyDevice>
void FlipGradient::forward_dev_impl(const MyDevice & dev, const vector<const Tensor*>& xs, Tensor& fx) const {
  fx.d = xs[0]->d;
  fx.v = xs[0]->v;
}

template<class MyDevice>
void FlipGradient::backward_dev_impl(const MyDevice & dev,
                                   const vector<const Tensor*>& xs,
                                   const Tensor& fx,
                                   const Tensor& dEdf,
                                   unsigned i,
                                   Tensor& dEdxi) const {
  dEdxi.tvec().device(*dev.edevice) = dEdf.tvec().unaryExpr(scalar_negative_op<float>());
}
DYNET_NODE_INST_DEV_IMPL(FlipGradient)
```
And this is from simd_funtors.h. 
```c
namespace dynet {
template<typename Scalar> struct scalar_negative_op {
  EIGEN_EMPTY_STRUCT_CTOR(scalar_negative_op)
  DYNET_DEVICE_FUNC inline const Scalar operator() (const Scalar& x) const {
    return -x;
  }
  template <typename Packet>
  DYNET_DEVICE_FUNC inline Packet packetOp(const Packet& x) const {
    using namespace Eigen::internal;
    return pnegate(x);
  }
};
}

namespace Eigen { namespace internal {
template<typename Scalar>
struct functor_traits<dynet::scalar_negative_op<Scalar> > {
  enum {
    Cost = NumTraits<Scalar>::MulCost * 2,
    PacketAccess = packet_traits<Scalar>::HasNegate
  };
};
} }
```

Currently I am writing a sample code to see if this works fine in actual training process. If I missed some part to be changed or got anything wrong, glad to have your feedback. 

Thanks,

**Comment by iamyoungjo** [comment]
I initiated a pull request. Thanks @neubig !

**Comment by neubig** [comment]
Resolved by #321
